Signaling Server API

simbleau opened the pull request:
Just looking for comments on a potential API design for a server API, re: #90 - 

One of the reasons I want a server API like this is so that I can have a game server that runs both the signalling server and the game world server logic in a bevy app.

But some other reasons we might want this:
- Reduce boilerplate: authoring your own signalling server in <50 lines
- Make the flow for signalling servers very familiar to the webrtc_socket
- Help simplify/share code with the future "matchbox cli"
- Help us produce more example signalling servers with different topologies

The general idea I am going for is to ditto the WebRtcSocket API with a new type like `SignallingConfig` and have 2 convenience functions: `config.full_mesh()` or `config.client_server()`. There's also the `SignallingServer::new_full_mesh(config)` similar to `WebRtcSocket::new_with_config(..)`.

Options in the config would be: `max_peers` (which would be 2 for one-to-one topology + full-mesh), `authentication` (query args, header authentication, etc.), `port`, `hostname`, `tls`, and things of that nature. Right now it's all a bit bare.

The return signature would be `config.full_mesh() -> (SignallingServer<FullMesh>, SignallingLoopFuture)` where `SignallingLoopFuture` is the server bind you need to await to run the signalling service, in a similar pattern to the webrtc socket today.

The reason for the generic `SignallingServer<FullMesh>` is to use a type state pattern to change the  API and prevent the API from being misused, sine the way the API would be different depending on the topology.

--

So, just to clarify, no changes to socket/protocol would be necessary. This would be a server API to help make the examples and with the goal of your own signalling server *much easier*.

johanhelsing commented:
Couple of questions:

1. What's the advantage of the type state pattern over simple structs `FullMeshSignallingServer` potentially implementing a `SignallingServer` trait?

2. Are we sure copying the socket architecture (struct + future) is a good idea? In the socket I did it mostly to fit into bevy and ggrs, which are not that async friendly... On the server you probably have more control over the architecture?

3. Should we consider providing lower level APIs exposing more implementation details? If we expose axum filters/endpoints/services/whatever-they're-called, that would probably allow people to add their own auth layers etc. and put the endpoint an appropriate path/port.

simbleau commented:
> What's the advantage of the type state pattern over simple structs FullMeshSignallingServer potentially implementing a SignallingServer trait?

There will be a large amount of shared logic, which could go into a trait, leading to a lot of trait logic duplication. There will also be nuanced differences between the two, like `server.host() -> Option<PeerId>` only existing on client/server. I am okay with separating them, but it will increase the amount of code a bit.

> Are we sure copying the socket architecture (struct + future) is a good idea? In the socket I did it mostly to fit into bevy and ggrs, which are not that async friendly... On the server you probably have more control over the architecture?

It's really not a great idea, but I plan to use the signalling server in Bevy as well. My idea is to couple a signalling server and game server together and make the game server run using Bevy. If you see a better architecture I'm all for it.

> Should we consider providing lower level APIs exposing more implementation details? If we expose axum filters/endpoints/services/whatever-they're-called, that would probably allow people to add their own auth layers etc. and put the endpoint an appropriate path/port.

That is on the TODOs

simbleau commented:
@johanhelsing I changed the API to a callback-style API. I added two tests for those callbacks at the bottom. There's also a router "map" function in the builder allowing the user access to inserting state/layering/middleware/etc.

Right now I haven't done the client/server signaling, so even if a ClientServer builder gets built it will behave like a FullMesh. That will change in the next few days.

I'd like you to take a look at it now because this will be a PR and I think it would be good if you gave it a peek early to do less work later.

garryod commented:
Really nice to see this taking shape. A few miscellaneous thoughts I've had are:

- Having the `FullMesh` & `ClientServer` implementations in their own modules would be really nice, we could hide them away in a `topologies` module making it super trivial to find what's unique to them.
- The more we can do as middleware the better, in my mind auth should be this if at all possible.

To join the discussion:

> There will be a large amount of shared logic, which could go into a trait, leading to a lot of trait logic duplication. There will also be nuanced differences between the two, like server.host() -> Option<PeerId> only existing on client/server. I am okay with separating them, but it will increase the amount of code a bit.

Another potential solution for this would be to use a block like `impl<T: Topology> SignallingSocket for T>`, but I reckon the type state pattern will play out better if we want to make things generic over more than one type (e.g. `SignallingServer<Topology, Authenticator>`)

> > Are we sure copying the socket architecture (struct + future) is a good idea? In the socket I did it mostly to fit into bevy and ggrs, which are not that async friendly... On the server you probably have more control over the architecture?
>
> It's really not a great idea, but I plan to use the signalling server in Bevy as well. My idea is to couple a signalling server and game server together and make the game server run using Bevy. If you see a better architecture I'm all for it.

I think sticking to async (likely struct + future) would be a good call, especially since the server itself is async. It's easy enough to wrap this in something (i.e. a bevy plugin) that makes it sync whilst if we were to go with a callback interface it would feel silly wrapping that as async again ending up with async -> callback -> async

simbleau commented:
Originally the PR took the same form as the socket, eg. (signaler, SignalerLoopFuture) 

Currently the direction I'm headed is a callback style API, where one can pass in lifecycle callbacks in the builder phase and then when the server is built, you just have one method, .serve which is an async "loop future." 

I think this is less surprising to the user and still very extensible.

Something I am purposefully removing is the matchmaking service from this API. It should be sufficiently easy to recreate it using callbacks.

johanhelsing reviewed (commented):
Sorry, I'm a bit pressed on time, so only got to take a very quick look at this, but I think the high-level public API makes sense.

Though I'm a bit confused about how the builder/type-state/callbacks would fit together if you plan to implement your own topology?

johanhelsing commented on matchbox_signalling_socket/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+name = "matchbox_signalling_socket"
+version = "0.5.0"
+authors = ["Johan Helsing <johanhelsing@gmail.com>", "Spencer C. Imbleau <spencer@imbleau.com>"]
+description = "Painless WebRTC peer-to-peer signalling socket"
```
Or maybe the crate name could have `server` in it?

simbleau replied:
Yeah, socket doesn't make sense anymore now. It did when the PR originally came up.

johanhelsing commented on matchbox_signalling_socket/src/signalling_socket/builder.rs:
```diff
@@ -0,0 +1,122 @@
+}
+
+impl SignalingServerBuilder<ClientServer> {
+    pub fn on_host_connected(mut self) -> Self {
```
So it's a setter for a callback, not the actual callback?

simbleau replied:
Yeah they are setters. It will look like on_peer_connect after. You're right it's missing parameters.

garryod commented:
> @johanhelsing @garryod Hey guys, sorry this is taking so long. This is turning out to be a nerd-sniping problem. I'm not the most brilliant with Pinnable futures and the async eco, but I'm slowly becoming an expert trying to implement this. I expect this may take another week to fully draft. I also apologize this is taking my time from other PRs and maintainer duties that are looking fantastic too.

Don't sweat it man, you're trying to solve a complex problem with some fairly hand wavy constraints. On the duties front... you're volunteering your time, we can't reasonably expect anything from you, we can only be grateful when that choose to give your time to this project. Let me know if there's anything I can do to help you with this one

simbleau commented:
@johanhelsing and @garryod this is now actually ready - callbacks work and the API is simple.

Still needs help (pitch in if you want!):
- Deploy to crates.io (@johanhelsing you will need to set this up)
- README changes - I just can't be bothered honestly
- More testing
- Recreate `matchbox_server` using `matchbox_signaling` - This can be done later, so it won't hold up this PR. But if you want to play with the new API, try this out. Note that matchmaking is not built in by default, so you'd have to do the matchmaking with callbacks.

Future improvement roadmap:
- Async callback support
- Add basic authentication on the builder (this may be doable by the user currently via middleware support)
- Have specific endpoints for a peer to connect to claim host. Currently the first peer to join becomes host, and future connections become clients.

johanhelsing commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,202 @@
+}
+
+impl SignalingServerBuilder<FullMesh, FullMeshCallbacks, FullMeshState> {
+    /// Set a callback triggered on signals.
+    pub fn on_signal<F>(mut self, callback: F) -> Self
+    where
+        F: Fn(JsonPeerRequest) + 'static,
+    {
+        self.callbacks.on_signal = Callback::from(callback);
+        self
+    }
```
What would be the use case of accessing the signal payload?

simbleau replied:
Not sure why someone would want this. Maybe they just want to log all the signals.

johanhelsing replied:
I think this is perhaps already covered by `trace` logging? Maybe remove it until somebody has a need for it? Think this would get rid of the need to clone the signal as well.

johanhelsing commented on matchbox_server/src/signaling.rs:
```diff
@@ -30,11 +30,6 @@ pub(crate) struct RequestedRoom {
     next: Option<usize>,
 }
 
-#[derive(serde::Deserialize, serde::Serialize)]
-pub(crate) struct QueryParam {
-    next: Option<usize>,
-}
-
```
Was this just unused previously?

simbleau replied:
It was unused previously, yep

johanhelsing reviewed (commented):
Okay, apologies for the mixed detail/quality of the review. Just jotted down my thoughts as I read through from top to bottom. Probably there are a few stupid questions and suggestions in there, but hopefully there are some useful ones as well :)

Overall, I think this looks really promising. There are some concepts I think I haven't fully wrapped my head around yet, but feels like it will make sense ;)

The examples look really neat, and the callbacks look like it would be possible to build other useful abstractions on top.

If we try to get it into 0.6: the intention is that it would be an alternative to `matchbox_server`, right? (just making sure my tutorial series is still making sense).

Also, like @garryod said, please don't feel pressured into working faster/harder on this than you have to, also very grateful for the work you do :)

garryod commented on matchbox_signaling/examples/server.rs:
```diff
@@ -0,0 +1,27 @@
+use std::net::Ipv4Addr;
+use tracing::info;
+
+#[tokio::main]
+async fn main() -> Result<(), matchbox_signaling::Error> {
+    // Setup logging
+    use tracing_subscriber::prelude::*;
+    tracing_subscriber::registry()
+        .with(
+            tracing_subscriber::EnvFilter::try_from_default_env()
+                .unwrap_or_else(|_| "debug".into()),
+        )
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let server = SignalingServer::client_server_builder((Ipv4Addr::UNSPECIFIED, 3536))
+        .on_host_connected(|id| info!("Host joined: {id:?}"))
+        .on_host_disconnected(|id| info!("Host left: {id:?}"))
+        .on_client_connected(|id| info!("Client joined: {id:?}"))
+        .on_client_disconnected(|id| info!("Client left: {id:?}"))
+        .on_signal(|s| info!("Signal: {s:?}"))
+        .cors()
+        .trace()
+        .build();
+    server.serve().await
+}
```
As above, it would be nice to move logging setup out of main for the sake of clarity

```suggestion
#[tokio::main]
async fn main() -> Result<(), matchbox_signaling::Error> {
    setup_logging();

    let server = SignalingServer::full_mesh_builder((Ipv4Addr::UNSPECIFIED, 3536))
        .on_peer_connected(|id| info!("Joined: {id:?}"))
        .on_peer_disconnected(|id| info!("Left: {id:?}"))
        .on_signal(|s| info!("Signal: {s:?}"))
        .cors()
        .trace()
        .build();
    server.serve().await
}

fn setup_logging() {
    use tracing_subscriber::prelude::*;
    tracing_subscriber::registry()
        .with(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| "debug".into()),
        )
        .with(tracing_subscriber::fmt::layer())
        .init();
}
```

garryod reviewed (commented):
I haven't had a whole lot of time to look in depth but this looks good from 3000ft. The few comments I have made are just places where I had to take pause when reading through. Will be gone for the afternoon but can have a proper look when I get back

simbleau commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+impl SignalingServerBuilder<ClientServer, ClientServerCallbacks, ClientServerState> {
```
I can do that, and yes

simbleau commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+impl SignalingServerBuilder<FullMesh, FullMeshCallbacks, FullMeshState> {
```
I can do that, and yes

simbleau commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,279 @@
+/// Signaling callbacks for client/server topologies
+#[derive(Default, Debug, Clone)]
+pub struct ClientServerCallbacks {
+    /// Triggered on peer requests to the signalling server
```
Yeah, noted

simbleau commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,202 @@
+            socket_addr: socket_addr.into(),
+            router: Router::new()
+                .route("/", get(ws_handler::<Cb, S>))
+                .route("/:path", get(ws_handler::<Cb, S>))
```
Yes, just not currently exposed to the user. Making a TODO for this

simbleau replied:
Now there's a shared callback on_connect() which passes you the extract of the connection (query params, IP addr, etc.) to do what you with with.

simbleau commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,279 @@
+
+impl ClientServerState {
+    /// Check if the host is available to take
+    pub fn is_host_available(&mut self) -> bool {
+        self.host.try_lock().as_ref().unwrap().is_none()
+    }
```
Sure I'll give this a go

simbleau commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    /// Send a message to a channel without blocking.
+    pub fn try_send(
+        &self,
```
It uses self.clients

simbleau commented on matchbox_signaling/src/topologies/full_mesh.rs:
```diff
@@ -0,0 +1,189 @@
+/// Signaling server state for full mesh topologies
+#[derive(Default, Debug, Clone)]
+pub struct FullMeshState {
+    pub peers: Arc<Mutex<HashMap<PeerId, UnboundedSender<Result<Message, axum::Error>>>>>,
```
Should be pub(crate). No one should extend this, probably.

The goal is just to provide 2 basic topologies by default: ClientServer and FullMesh

simbleau replied:
I take it back, it needs to be public since it goes into the generic for the builder and that can't leak (enforced by compiler)

simbleau commented:
@johanhelsing yes currently the intention is to provide this crate as an alternative to the server. After this looks good then we can replace the code within matchbox_server to use this API.

@garryod I also forgot to mention, I'd really like if this signaling server could be a Bevy Plugin! It's very likely that people would want to pair a signaling server and Host peer together, to act as a server.

simbleau commented on matchbox_protocol/src/lib.rs:
```diff
@@ -10,14 +10,14 @@ use uuid::Uuid;
 
 /// Requests go from peer to signalling server
-#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
```
I think it was to send in a callback, will double check

simbleau commented on matchbox_signaling/tests/client_server.rs:
```diff
@@ -0,0 +1,153 @@
+
+    #[tokio::test]
+    async fn ws_on_host_connect_callback() {
+        let success = Arc::new(AtomicBool::new(false));
```
Yeah it would be nice if these tests had less boilerplate, but I don't see the need to take this action item. If you want, you may :)

garryod commented:
> @garryod I also forgot to mention, I'd really like if this signaling server could be a Bevy Plugin! It's very likely that people would want to pair a signaling server and Host peer together, to act as a server.

I think you must've mentioned it somewhere else, or we had the same idea :smile: but I totally agree, this should definitely be part of the bevy plugin (either behind a feature or a cfg `native`)

simbleau commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,202 @@
+
+    /// Create a [`SignalingServer`].
+    ///
+    /// # Panics
+    /// This method will panic if the socket address requested cannot be bound.
```
axum::Server::bind

simbleau commented:
@garryod  @johanhelsing It's complete! Moving this out of Draft.

Action Items:
- @johanhelsing I need you to add the CI to deploy this to crates.io

I'll setup separate issues for the following (Not in this PR):
- Recreate matchbox_server using matchbox_signaling
- Async callback support
- Add basic authentication
- Bevy plugin integration

simbleau commented on examples/simple/src/main.rs:
```diff
@@ -21,7 +21,7 @@ async fn main() {
         .with(
             tracing_subscriber::EnvFilter::try_from_default_env()
-                .unwrap_or_else(|_| "simple=info,matchbox_socket=info".into()),
+                .unwrap_or_else(|_| "simple_example=info,matchbox_socket=info".into()),
```
this was a typo I added to this PR

garryod commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,296 @@
+            info!("{peer_uuid:?} -> {event_text}");
+        };
+
+        // TODO: Make some real way to validate hosts, authentication?
+        // Currently, the first person to connect becomes host.
```
I think this behaviour is fine, any kind of auth can be added as middleware.

simbleau replied:
Working on auth now

garryod commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,139 @@
+        self
+    }
+
+    /// Change the topology.
+    pub fn topology(mut self, topology: Topology) -> Self {
+        self.topology = topology;
+        self
+    }
```
I might be missing something very obvious here, but how does this work when we have the type state pattern locking us to a given Topology?

simbleau replied:
In my mind, originally this was useful. I don't know if it's still useful.

My original thought was, if you own your own topology, like OneToOne, you'd still use FullMeshState and FullMeshCallbacks likely. So changing the topology would make a Builder<OneToOne, FullMeshCallbacks, FullMeshState> if you started with a full_mesh_builder and changed the topology to OneToOne.

I do agree this seems useless in practice... it should be removed. Creating a new topology is an endeavor and something not likely someone would do, as well this would probably confuse them.

garryod replied:
> started with a full_mesh_builder and changed the topology to OneToOne

I don't believe this is possible as `topology` must be of type `Topology`, which is already bound to the current topology. This could be fixed by having the function create a new correctly typed builder, but I don't think it's really worth the effort when anyone creating a Topology can use an extension trait to create a `new_xyz` function on the builder

garryod commented on matchbox_signaling/src/signaling_server/builder.rs:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    /// Modify the router with a mutable closure. This is where one may apply middleware or other
+    /// layers to the Router.
+    pub fn middleware(mut self, mut alter: impl FnMut(&mut Router)) -> Self {
+        alter(&mut self.router);
+        self
+    }
```
Is this perhaps too permissive and ergo a possible foot gun? I feel like we probably want it to be an alias to `Router::layer`

simbleau replied:
I'd rather start permissive and if we get issues/people are using it the wrong way then we can scope it down. Open to opinions

garryod replied:
That's fair, I suppose it's kind of an expert user feature anyway

garryod commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,296 @@
+        let (ws_sender, mut ws_receiver) = ws.split();
+        let sender = spawn_sender_task(ws_sender);
+
+        // Generate a UUID for the user
+        let peer_uuid = uuid::Uuid::new_v4().into();
```
Should this be offered as a callback? Might be nice if somebody wanted to add the ability to resume sessions or use meaningful uuids

simbleau replied:
Yea, that's fair. Will do.

garryod commented on matchbox_signaling/src/signaling_server/handlers.rs:
```diff
@@ -0,0 +1,73 @@
+        query_params,
+    };
+
+    // Lifecycle event: On Upgrade
+    let allow_cxn = shared_callbacks.on_upgrade.emit(extract);
+
+    // Finalize the upgrade process by returning upgrade callback to client
+    if allow_cxn {
```
I would rather this wasn't abbreviated

```suggestion
    // Lifecycle event: On Upgrade
    let allow_connection = shared_callbacks.on_upgrade.emit(extract);

    // Finalize the upgrade process by returning upgrade callback to client
    if allow_connection {
```

garryod commented on matchbox_signaling/src/topologies/full_mesh.rs:
```diff
@@ -0,0 +1,195 @@
+/// Signaling server state for full mesh topologies
+#[derive(Default, Debug, Clone)]
+pub struct FullMeshState {
+    pub peers: StateObj<HashMap<PeerId, SignalingChannel>>,
```
Do we want to make this `pub` or would `pub(crate)` be more appropriate?

garryod reviewed (commented):
Looking good, just a few little nit-picks

simbleau commented:
@garryod Tackled all those feedback items.

Otherwise, I actually implemented an entire working Header based authentication layer with Basic auth and then realized Tungstenite doesn't allow one to send headers on connection... disappointment.

So I did the next best thing, which is to expose headers (incase Tungstenite changes their mind in the future) and added a token based (base64) query parameter auth method as an interim solution for those wanting auth.

simbleau commented:
Ok, the herculean effort is complete, I'm quite happy with this. Last chance for feedback before it gets merged :)

garryod commented:
> Couldn't find any alternatives? 

Having had a brief look it seems like `async_tungstenite::tokio::client_async` might allow us to do this as it can take a request with a header. Though it seems to need a pre-existing stream so may only be useful for promotion instead of straight to WS

garryod commented:
Scratch that, `async_tungstenite::tokio::connect_async` takes a `tungstenite::handshake::client::Request` which can be built like so:

```rust
let request = Request::builder()
            .uri(room_url)
            .header("auth", "some_value")
            .body(())
            .unwrap();
```

garryod commented:
That being said, it may be best we remove auth for the sake of this PR and include it in a follow up as it's likely to warrant some discussion

simbleau commented:
> Scratch that, `async_tungstenite::tokio::connect_async` takes a `tungstenite::handshake::client::Request` which can be built like so:
> 
> ```rust
> let request = Request::builder()
>             .uri(room_url)
>             .header("auth", "some_value")
>             .body(())
>             .unwrap();
> ```

![image](https://user-images.githubusercontent.com/48108917/227802431-01b3cf2d-9dd3-431e-8df9-6ca78aaa7dc0.png)

simbleau commented:
Ok, basic authentication works now. However, matchbox_socket doesn't have any way to use basic auth :P So I'll do a separate PR for that

garryod commented:
Aha, nice one, might be better to hold back the `matchbox_signalling` implementation of it too so they can be implemented / discussed as one

simbleau commented:
Well, it's done and completed with testing. But yeah, I'll hold off for further auth layers like Bearer authentication.

garryod commented:
Sounds good, would be real sweet to get it out for `0.6`

johanhelsing commented on matchbox_signaling/examples/client_server.rs:
```diff
@@ -0,0 +1,34 @@
+
+    let server = SignalingServer::client_server_builder((Ipv4Addr::UNSPECIFIED, 3536))
+        .on_connection_request(|cxn| {
+            info!("Connecting: {cxn:?}");
```
I also find the `cxn` abbreviation a bit weird. Better to spell it out, since the lambda doesn't include a type annotation?

johanhelsing commented on matchbox_signaling/src/signaling_server/handlers.rs:
```diff
@@ -0,0 +1,114 @@
+
+    // Finalize the upgrade process by returning upgrade callback to client
+    // Generate an ID for the peer
+    let peer_id = uuid::Uuid::new_v4().into();
```
There was some discussion of making it possible to override this (e.g. on reconnections/meaningful uuids), was this left for a follow up PR?

simbleau replied:
There's callbacks for id assignment, but no way to not send an ID. That can be future work if someone wants it.

johanhelsing commented on matchbox_signaling/src/signaling_server/mod.rs:
```diff
@@ -0,0 +1,40 @@
+#[derive(Default, Debug, Clone)]
+pub struct NoOpCallouts {}
+impl SignalingCallbacks for NoOpCallouts {}
+#[allow(unsafe_code)]
+unsafe impl Send for NoOpCallouts {}
+#[allow(unsafe_code)]
+unsafe impl Sync for NoOpCallouts {}
```
These `#[allow(unsafe_code)]`s still need some explanation.

garryod replied:
As far as I can tell, it all derives from the use of `Rc<dyn Fn(In) -> Out>` in `Callback`. @simbleau Is there a reason you've chosen to go this way instead of an `Arc<Mutex<dyn Fn(In) -> Out>>` or likewise?

johanhelsing commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,282 @@
+                            state.try_send_to_host(event)
+                        }
+                    } {
+                        error!("error sending: {:?}", e);
```
```suggestion
                        error!("error sending signal event: {e:?}");
```

johanhelsing commented on matchbox_signaling/src/topologies/client_server.rs:
```diff
@@ -0,0 +1,282 @@
+                }
+            };
+
+            match request {
```
This block will probably be the same in 98% of cases, is there a way we could deduplicate it? Or provide helpers that make it shorter?

johanhelsing commented on matchbox_signaling/src/topologies/full_mesh.rs:
```diff
@@ -0,0 +1,183 @@
+    pub(crate) on_peer_disconnected: Callback<PeerId>,
+}
+impl SignalingCallbacks for FullMeshCallbacks {}
+#[allow(unsafe_code)]
+unsafe impl Send for FullMeshCallbacks {}
+#[allow(unsafe_code)]
+unsafe impl Sync for FullMeshCallbacks {}
```
More unexplained unsafe

johanhelsing reviewed (changes requested):
I'm a bit pressed on time again (generally my open-source days are thu-sun).

As I see it the only things that really needs to be done is explain the unsafe usage, and I think I'd also want to move the authentication to another PR, especially since we want to get this PR into 0.6.

simbleau commented:
Auth is removed, all comments seem addressed now. No unsafe, etc.

Merging once pipeline passes.
